File: Homework04/Zipcode.cs
using System;
using System.Collections.Generic;
using System.IO;

public class Zipcode 
{
    public int RecordNumber { get; set; }
    public int ZipCode { get; set; }
    public string ZipCodeType { get; set; }
    public string City { get; set; }
    public string State { get; set; }
    public string LocationType { get; set; }
    public double Lat { get; set; }
    public double Long { get; set; }
    public double Xaxis { get; set; }
    public double Yaxis { get; set; }
    public double Zaxis { get; set; }
    public string WorldRegion { get; set; }
    public string Country { get; set; }
    public string LocationText { get; set; }
    public string Location { get; set; }
    public bool Decommissioned { get; set; }
    public int TaxReturnsFiled { get; set; }
    public int EstimatedPopulation { get; set; }
    public int TotalWages { get; set; }

    public static List<Zipcode> readZipCodes(string file) 
    {
        List<Zipcode> zipcodes = new List<Zipcode>();
        string[] lines = File.ReadAllLines(file);
        foreach (string line in lines) 
        {
            string[] values = line.Split('\t');
            Zipcode zipcode = new Zipcode 
            {
                RecordNumber = values.Length > 0 && int.TryParse(values[0], out int recordNumber) ? recordNumber : 0,
                ZipCode = values.Length > 1 && int.TryParse(values[1], out int zipCode) ? zipCode : 0,
                ZipCodeType = values.Length > 2 ? values[2] : "",
                City = values.Length > 3 ? values[3] : "",
                State = values.Length > 4 ? values[4] : "",
                LocationType = values.Length > 5 ? values[5] : "",
                Lat = values.Length > 6 && double.TryParse(values[6], out double lat) ? lat : 0.0,
                Long = values.Length > 7 && double.TryParse(values[7], out double lon) ? lon : 0.0,
                Xaxis = values.Length > 8 && double.TryParse(values[8], out double x) ? x : 0.0,
                Yaxis = values.Length > 9 && double.TryParse(values[9], out double y) ? y : 0.0,
                Zaxis = values.Length > 10 && double.TryParse(values[10], out double z) ? z : 0.0,
                WorldRegion = values.Length > 11 ? values[11] : "",
                Country = values.Length > 12 ? values[12] : "",
                LocationText = values.Length > 13 ? values[13] : "",
                Location = values.Length > 14 ? values[14] : "",
                Decommissioned = values.Length > 15 && bool.TryParse(values[15], out bool decommissioned) ? decommissioned : false,
                TaxReturnsFiled = values.Length > 16 && int.TryParse(values[16], out int taxReturns) ? taxReturns : 0,
                EstimatedPopulation = values.Length > 17 && int.TryParse(values[17], out int population) ? population : 0,
                TotalWages = values.Length > 18 && int.TryParse(values[18], out int wages) ? wages : 0,
            };
            zipcodes.Add(zipcode);
        }
        return zipcodes;
    }
}

File: Homework05/zipcode.py


class Zipcode:
    def __init__(self, record_number=None, zipcode=None, zipcode_type=None, city=None, state=None,
                 location_type=None, lat=None, long=None, xaxis=None, yaxis=None, zaxis=None,
                 world_region=None, country=None, location_text=None, location=None, decommissioned=None,
                 tax_returns_filed=None, estimated_population=None, total_wages=None, notes=None):
        self.record_number = record_number
        self.zipcode = zipcode
        self.zipcode_type = zipcode_type
        self.city = city
        self.state = state
        self.location_type = location_type
        self.lat = lat
        self.long = long
        self.xaxis = xaxis
        self.yaxis = yaxis
        self.zaxis = zaxis
        self.world_region = world_region
        self.country = country
        self.location_text = location_text
        self.location = location
        self.decommissioned = decommissioned
        self.tax_returns_filed = tax_returns_filed
        self.estimated_population = estimated_population
        self.total_wages = total_wages
        self.notes = notes

    def __repr__(self):
        return f"Zipcode({self.zipcode})"

    def read_zipcodes(file_path):
        zipcodes = []
        with open(file_path, 'r') as file:
            headers = file.readline().strip().split('\t')

            for line in file:
                values = line.strip().split('\t')
                values = [val if val != '' else None for val in values]
                zipcode_data = dict(zip(headers, values))
            
                zipcode = Zipcode(
                    record_number=zipcode_data.get("RecordNumber"),
                    zipcode=zipcode_data.get("Zipcode"),
                    zipcode_type=zipcode_data.get("ZipCodeType"),
                    city=zipcode_data.get("City"),
                    state=zipcode_data.get("State"),
                    location_type=zipcode_data.get("LocationType"),
                    lat=float(zipcode_data["Lat"]) if zipcode_data.get("Lat") else None,
                    long=float(zipcode_data["Long"]) if zipcode_data.get("Long") else None,
                    xaxis=float(zipcode_data["Xaxis"]) if zipcode_data.get("Xaxis") else None,
                    yaxis=float(zipcode_data["Yaxis"]) if zipcode_data.get("Yaxis") else None,
                    zaxis=float(zipcode_data["Zaxis"]) if zipcode_data.get("Zaxis") else None,
                    world_region=zipcode_data.get("WorldRegion"),
                    country=zipcode_data.get("Country"),
                    location_text=zipcode_data.get("LocationText"),
                    location=zipcode_data.get("Location"),
                    decommissioned=zipcode_data.get("Decommisioned") == 'true',
                    tax_returns_filed=int(zipcode_data["TaxReturnsFiled"]) if zipcode_data.get("TaxReturnsFiled") else None,
                    estimated_population=int(zipcode_data["EstimatedPopulation"]) if zipcode_data.get("EstimatedPopulation") else None,
                    total_wages=float(zipcode_data["TotalWages"]) if zipcode_data.get("TotalWages") else None,
                    notes=zipcode_data.get("Notes")
                )
                zipcodes.append(zipcode)
        return zipcodes
